src/system/SystemLayer.cpp

```diff
@@ -303,6 +303,49 @@ Error Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, vo
     return lReturn;
 }
 
+/**
+ * @brief
+ *   This method extends a one-shot timer.
+ *
+ *   @note
+ *       Only a single timer is allowed to be started with the same @a aComplete and @a aAppState
+ *       arguments. If called with @a aComplete and @a aAppState identical to an existing timer,
+ *       the currently-running timer will be extended for a duration of specified period.
+ *
+ *   @param[in]  aMilliseconds Extension time in milliseconds.
+ *   @param[in]  aComplete     A pointer to the function called when timer expires.
+ *   @param[in]  aAppState     A pointer to the application state object used when timer expires.
+ *
+ *   @return CHIP_SYSTEM_NO_ERROR On success.
+ *   @return CHIP_SYSTEM_ERROR_NO_MEMORY If a timer cannot be allocated.
+ *   @return CHIP_SYSTEM_ERROR_UNEXPECTED_STATE If System Layer is not initialized.
+ *   @return Other Value indicating timer failed to start or extend.
+ *
+ */
+Error Layer::ExtendTimer(uint32_t aMilliseconds, Layer::TimerCompleteFunct aOnComplete, void * aAppState)
+{
+    VerifyOrReturnError(this->State() == kLayerState_Initialized, CHIP_SYSTEM_ERROR_UNEXPECTED_STATE);
+
+    Timer * lTimer = nullptr;
+    for (size_t i = 0; i < Timer::sPool.Size(); ++i)
+    {
+        Timer * timer = Timer::sPool.Get(*this, i);
+
+        if (timer != nullptr && timer->OnComplete == aOnComplete && timer->AppState == aAppState)
```

Comment:
This looks a bit dodgy.  Is there anything in the timer API that prevents duplicate registration with the same callback function and `aAppState`?  Should such a thing be illegal?  What if it is useful to register multiple timers with the same callback?

It seems like the timer API needs to return a handle to the created timer so that this can be passed back to `ExtendTimer`.

Reply:
We always call this->CancelTimer(aComplete, aAppState) within StartTimer before start an new timer to prevent duplicate registration. ExtendTimer is calling StartTimer to extend a timer